### ISSUE TITLE
using "gather_for_metrics"

### ISSUE THREAD

**Issue by lovodkin93** [huggingface/accelerate#930]
### System Info

```Shell
- `Accelerate` version: 0.15.0.dev0
- Platform: Linux-3.10.0-1160.71.1.el7.x86_64-x86_64-with-glibc2.10
- Python version: 3.8.5
- Numpy version: 1.23.4
- PyTorch version (GPU?): 1.13.0 (True)
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: fp16
        - use_cpu: False
        - dynamo_backend: NO
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: 0,1
        - main_process_ip: None
        - main_process_port: None
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - deepspeed_config: {}
        - fsdp_config: {}
        - megatron_lm_config: {}
        - downcast_bf16: no
        - tpu_name: None
        - tpu_zone: None
        - command_file: None
        - commands: None
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Hello,
I am trying to employ the `gather_for_metrics` function to do evaluation on all the devset.
I am following the example in this link:
https://github.com/huggingface/accelerate/blob/main/examples/by_feature/automatic_gradient_accumulation.py#L196-L204
The only difference is the fact that I do not use `metric` as I am using my own metric.
In stead, I opt for aggregating all the scores and then calculating the final score.
This is a snippet of my code with the relevant part:
```
        generated_outputs, precision_scores, recall_scores, final_rewards = [], [], [], []
        for i, (input_ids, attention_mask, global_attention_mask, jsonl_ids) in enumerate(tqdm(self.val_dataloader)):
            with torch.no_grad():
                input_ids, attention_mask, global_attention_mask = self.add_control_code(input_ids, attention_mask, global_attention_mask)
                rollouts = self.policy.sample(input_ids=input_ids, attention_mask=attention_mask, top_p=self.params.top_p, global_attention_mask=global_attention_mask)
                
                forward_inputs = {'input_texts_input_ids': rollouts['input_doc/input_ids'][:, 1:],
                                  'input_texts_attention_mask': rollouts['input_doc/mask'][:, 1:],
                                  'input_texts_global_attention_mask': rollouts['input_doc/global_mask'][:, 1:],
                                  'generated_reductions_input_ids': rollouts['output_reduction/input_ids'],
                                  'generated_reductions_attention_mask': rollouts['output_reduction/mask']}
                ref_logprobs = self.ref_policy.forward_pass(**forward_inputs)['generated_reduction/log_prob'] 

                input_doc = [self.policy.tokenizer.decode(p, clean_up_tokenization_spaces=True) for p in rollouts['input_doc/input_ids'][:, 1:]] 
    
                output_reduction = rollouts['output_reduction/text']
                score = self.score_model.get_reward(input_docs=input_doc, jsonl_ids=jsonl_ids, output_reductions=output_reduction, epoch=f'step{step}_eval{i}', jsonl_path=self.params.dataset_val)
                precision_score, recall_score, final_reward = score["precision_score"], score["recall_score"], score["final_reward"]


                output_reduction, precision_score, recall_score, final_reward  = self.accelerator.gather_for_metrics((output_reduction, precision_score, recall_score, final_reward))
                
                generated_outputs.extend(rollouts['output_reduction/text'])
                precision_scores.extend(precision_score)
                recall_scores.extend(recall_score)
                final_rewards.extend(final_reward)
```
The problem is that in the end, `generated_outputs`, `precision_scores`, `recall_scores`, `final_rewards` are all empty lists.
What am I doing wrong?
And a followup question - how do I reset the aggregation after calculating the final score for the current epoch? (As in the example, calling `metric.compute()` resets the aggregation, but in my case I do not use metric).
Thanks! 

### Expected behavior

```Shell
having all the scores and generated outputs in `generated_outputs`, `precision_scores`, `recall_scores`, `final_rewards`
```

**Comment by sgugger** [comment]
cc @muellerzr

**Comment by muellerzr** [comment]
@lovodkin93 a few questions on this:

- Are `score["precision_score"], score["recall_score"], score["final_reward"]` all `torch.tensor`'s?
- Does the code work on a single GPU?
- Is any part of your previous code as it goes through and back expected to have *all* of the actual predictions rather than the ones from a single worker/node?

Thanks!

**Comment by varadhbhatnagar** [comment]
@lovodkin93 From the documentation, it seems that `gather_for_metrics()` only works on `tensors` and not on other types.
Let me know if you have found a way to do this for variables with other types which may be part of the batch.
https://huggingface.co/docs/accelerate/v0.19.0/en/package_reference/accelerator#accelerate.Accelerator.gather_for_metrics
